File: src/net/machinemuse/numina/mouse/MouseEventHandler.scala
package net.machinemuse.numina.mouse

import net.minecraftforge.event.ForgeSubscribe
import net.minecraftforge.client.event.MouseEvent
import net.minecraft.client.Minecraft
import net.machinemuse.numina.item.ModeChangingItem
import net.machinemuse.numina.scala.OptionCast

/**
 * Author: MachineMuse (Claire Semple)
 * Created: 12:40 PM, 9/5/13
 */
object MouseEventHandler {
  @ForgeSubscribe def onMouseEvent(e: MouseEvent) {
    if (e.dwheel != 0) {
      for {
        p <- Option(Minecraft.getMinecraft.thePlayer)
        stack <- Option(p.getCurrentEquippedItem)
        item <- OptionCast[ModeChangingItem](stack.getItem)
        if p.isSneaking
      } {
        item.cycleMode(stack, e.dwheel / 120)
        e.setCanceled(true)
      }
    }
  }
}

File: src/main/java/net/machinemuse/numina/mouse/MouseEventHandler.java
package net.machinemuse.numina.mouse;

import cpw.mods.fml.common.eventhandler.SubscribeEvent;
import net.machinemuse.numina.item.IModeChangingItem;
import net.minecraft.client.Minecraft;
import net.minecraft.client.entity.EntityClientPlayerMP;
import net.minecraft.item.ItemStack;
import net.minecraftforge.client.event.MouseEvent;

/**
 * Author: MachineMuse (Claire Semple)
 * Created: 12:40 PM, 9/5/13
 *
 * Ported to Java by lehjr on 10/22/16.
 */
public final class MouseEventHandler {
    @SubscribeEvent
    public void onMouseEvent(MouseEvent e) {
        if (e.dwheel != 0) {
            EntityClientPlayerMP player = Minecraft.getMinecraft().thePlayer;
            ItemStack stack = player.getCurrentEquippedItem();
            if (stack != null && stack.getItem() instanceof IModeChangingItem) {
                IModeChangingItem item = (IModeChangingItem) stack.getItem();
                if (player.isSneaking()) {
                    item.cycleMode(stack, player, e.dwheel / 120);
                    e.setCanceled(true);
                }
            }
        }
    }
}